### PR TITLE
Update rules nodejs

### PR DESCRIPTION
**Pull request by josephperrott**
See individual commits.

### CODE REVIEW

File: WORKSPACE [hunk 1]
```diff
@@ -61,22 +53,15 @@ yarn_install(
     ],
     package_json = "//:package.json",
     quiet = False,
+    strict_visibility = False,
```
Comment (jelbourn) [inline WORKSPACE:56]: Could we add a comment here that explains why this is set to false?

File: tools/highlight-files/BUILD.bazel [hunk 2]
```diff
@@ -21,5 +22,10 @@ nodejs_binary(
         "@npm//fs-extra",
         "@npm//highlight.js",
     ],
+    templated_args = [
+            # TODO(josephperrott): update dependency usages to no longer need bazel patch module resolver
+            # See: https://github.com/bazelbuild/rules_nodejs/wiki#--bazel_patch_module_resolver-now-defaults-to-false-2324
```
Comment (jelbourn) [inline tools/highlight-files/BUILD.bazel:27]: Is it intentional to only add this comment for a couple of instances of this flag?

File: tools/postinstall/apply-patches.js [hunk 3]
```diff
@@ -87,7 +87,7 @@ function applyPatches() {
     const hasFlatModuleBundle = fs.existsSync(filePath.replace('.d.ts', '.metadata.json'));
     if ((filePath.includes('node_modules/') || !hasFlatModuleBundle) && $1`,
       'node_modules/@angular/compiler-cli/src/transformers/compiler_host.js');
-  applyPatch(path.join(__dirname, './flat_module_factory_resolution.patch'));
+//  applyPatch(path.join(__dirname, './flat_module_factory_resolution.patch'));
```
Comment (jelbourn) [inline tools/postinstall/apply-patches.js:90]: Commented lines?

File: test/bazel-karma-local-config.js [hunk 4]
```diff
@@ -1,14 +1,14 @@
 /**
- * Karma configuration that is used by Bazel karma_web_test targets which do not
+ * concatjs configuration that is used by Bazel concatjs_web_test targets which do not
```
Comment (jelbourn) [inline test/bazel-karma-local-config.js:2]: Were the changes in this file intentional? They look like they might be an overzealous find & replace.

File: packages.bzl [hunk 5]
```diff
@@ -9,6 +9,7 @@ VERSION_PLACEHOLDER_REPLACEMENTS = {
     "0.0.0-MDC": MDC_PACKAGE_VERSION,
     "0.0.0-NG": ANGULAR_PACKAGE_VERSION,
     "0.0.0-TSLIB": TSLIB_PACKAGE_VERSION,
+    "0.0.0-PLACEHOLDER": "{BUILD_SCM_VERSION}",
```
Comment (jelbourn) [inline packages.bzl:12]: Could we add a comment for what this placeholder is used for? The others are already pretty self-explanatory

**Review by crisbeto — APPROVED** [review]
LGTM once Jeremy's comments have been addressed.

**Comment by josephperrott** [comment]
@crisbeto and @jelbourn I meant to create this as a draft PR initially, but mistakenly made it as ready for review.

I have cleaned it up and have everything passing.  Go ahead and take a look again please.

File: test/bazel-karma-local-config.js [hunk 6]
```diff
@@ -3,12 +3,12 @@
  * want to launch any browser and just enable manual browser debugging.
  */
 
-const bazelKarma = require('@bazel/karma');
+const bazelKarma = require('@bazel/concatjs');
```
Comment (jelbourn) [inline test/bazel-karma-local-config.js:6]: I find it really strange that they changed these rules to live under a package named for the bundler 😕

**Review by jelbourn — APPROVED** [review]
LGTM

File: test/bazel-karma-local-config.js (hunk 6 above)
Comment (josephperrott) [reply test/bazel-karma-local-config.js:6]: Yeah, my guess is that its about consolidation overall: https://github.com/bazelbuild/rules_nodejs/wiki#introducing-bazelconcatjs

File: package.json [hunk 7]
```diff
@@ -71,26 +71,75 @@
     "zone.js": "~0.11.3"
   },
   "devDependencies": {
+    "diff": "^3.5.0",
+    "reflect-metadata": "^0.1.3",
+    "@material/animation": "^10.0.0",
+    "@material/auto-init": "^10.0.0",
```
Comment (crisbeto) [inline package.json:77]: Shouldn't these match the version of `material-components-web`?

File: src/cdk/schematics/ng-add/index.spec.ts [hunk 8]
```diff
@@ -4,6 +4,10 @@ import {COLLECTION_PATH} from '../index.spec';
 import {createTestApp, getFileContent} from '../testing';
 import {addPackageToPackageJson} from './package-config';
 
+interface PackageJson {
+  dependencies: {[key: string]: string};
```
Comment (crisbeto) [inline src/cdk/schematics/ng-add/index.spec.ts:8]: Can be simplified to `Record<string, string>`. Also this is being repeated in a few test files, should we reuse it?

**Review by crisbeto — APPROVED** [review]
Still LGTM with a couple of nits.

File: package.json (hunk 7 above)
Comment (josephperrott) [reply package.json:77]: Done, I didn't realize that the canary versions set different versions on the deps they had.

File: src/cdk/schematics/ng-add/index.spec.ts (hunk 8 above)
Comment (josephperrott) [reply src/cdk/schematics/ng-add/index.spec.ts]: I am not sure the best option for a location for us to put this common interface. Any ideas?

Comment (crisbeto) [reply src/cdk/schematics/ng-add/index.spec.ts]: I don't see a good place to add it. Let's just leave it as it is for now.

File: .nvmrc [hunk 9]
```diff
@@ -0,0 +1 @@
+12.14.1
```
Comment (Splaktar) [inline .nvmrc:1]: Is there any reason this can't just be `12`? As is, it won't let me commit when I'm using NodeJS `12.21.0`.

Comment (josephperrott) [reply .nvmrc:1]: The intention is that by using the exact same version here and on CI, there are not any unexpected differences between our usages.

**Pull request merged**